source/fe/fe.cc

```diff
@@ -340,12 +350,86 @@ FiniteElement<dim, spacedim>::get_restriction_matrix(
            "Restriction matrices are only available for refined cells!"));
   AssertIndexRange(
     child, GeometryInfo<dim>::n_children(RefinementCase<dim>(refinement_case)));
-  // we use refinement_case-1 here. the -1 takes care of the origin of the
-  // vector, as for RefinementCase<dim>::no_refinement (=0) there is no data
+
+  // We use refinement_case-1 here. The -1 takes care of the origin of the
+  // vector, as for RefinementCase::no_refinement (= 0) there is no data
   // available and so the vector indices are shifted
-  Assert(restriction[refinement_case - 1][child].n() == this->n_dofs_per_cell(),
-         ExcProjectionVoid());
-  return restriction[refinement_case - 1][child];
+  const auto &restriction_matrix = restriction[refinement_case - 1][child];
+
+  restriction_matrix.ensure_initialized([&]() {
+    const auto matrix = construct_restriction_matrix(child, refinement_case);
+    Assert(matrix.has_value(), ExcProjectionVoid());
+    return matrix.value();
+  });
+
+  Assert(restriction_matrix.has_value(), ExcProjectionVoid());
+  return restriction_matrix.value();
+}
+
+
+
+template <int dim, int spacedim>
+bool
+FiniteElement<dim, spacedim>::restriction_is_implemented() const
+{
+  for (unsigned int ref_case = RefinementCase<dim>::cut_x;
+       ref_case < RefinementCase<dim>::isotropic_refinement + 1;
+       ++ref_case)
+    for (unsigned int c = 0;
+         c < GeometryInfo<dim>::n_children(RefinementCase<dim>(ref_case));
+         ++c)
+      {
+        // True to initialize the restriction matrix, but fail gracefully
+        // and return false if something went wrong:
+        try
+          {
+            restriction[ref_case - 1][c].ensure_initialized([&]() {
+              const auto matrix =
+                construct_restriction_matrix(c, RefinementCase<dim>(ref_case));
+              if (!matrix.has_value())
+                throw ExcProjectionVoid();
+              return matrix.value();
+            });
```

Comment:
I would prefer if in the body here you just call `get_restriction_matrix()`, since you then don't have to duplicate the code for the initialization.

Regardless, this is of course quite expensive because you are requesting all of the matrices. I'm in favor of deprecating the function.

Reply:
The problem is that `get_restriction_matrix()` will assert that we can indeed return a restriction matrix. Meaning that the `restriction_is_implemented()` function can only ever return `true`, or abort the program. So the little dance here makes the function work as advertised.

I am happy to deprecate the function as well.

Reply:
What if you changed the `get_restriction_matrix()` function to use `AssertThrow()`? Then you could call it and if it throws, catch that and do `return false;`.

Reply:
@bangerth As I said - I would like to not use the exception mechanism in this case because this would require that derived FE classes cooperate and also `AssertThrow` if they have an issue (or if a variant is not implemented). The `std::optional` would make the intend of the `construct_*` clear.

That said, I am leaning towards not overengineering this: Remove the `std::optional` from the `construct_*` methods and simply `Assert` that everything goes well. And deprecate the `*_is_implemented()` methods and keep the broken behavior.

Reply:
Yes, perhaps not worth fretting over if we deprecate the function anyway.